// @fabiospampinato/bump/dist/providers/release/github.js
"use strict";
/* IMPORT */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var chalk_1 = require("chalk");
var fs = require("fs");
var globby = require("globby");
var mime = require("mime-types");
var octokit = require("@octokit/rest");
var opn = require("opn");
var path = require("path");
var username = require("git-username");
var config_1 = require("../../config");
var utils_1 = require("../../utils");
var file_1 = require("../changelog/file");
/* COMMIT */
var GitHub = {
    do: function (repoPath, version) {
        return __awaiter(this, void 0, void 0, function () {
            var cwd, github, owner, repo, tag, release, _a, _b, _c, filePaths, _i, filePaths_1, filePath, e_1;
            return __generator(this, function (_d) {
                switch (_d.label) {
                    case 0:
                        if (!config_1.default.release.github.token)
                            return [2 /*return*/, utils_1.default.exit('[release] Missing GitHub token')];
                        cwd = repoPath, github = new octokit({
                            auth: config_1.default.release.github.token
                        });
                        _d.label = 1;
                    case 1:
                        _d.trys.push([1, 9, , 10]);
                        owner = config_1.default.release.github.owner || username({ cwd: cwd }), repo = config_1.default.release.github.repo || path.basename(cwd), tag = utils_1.default.template.render(config_1.default.tag.name, { version: version });
                        _b = (_a = github.repos).createRelease;
                        _c = {
                            owner: owner,
                            repo: repo,
                            draft: config_1.default.release.github.draft,
                            prerelease: config_1.default.release.github.prerelease,
                            name: tag,
                            tag_name: tag
                        };
                        return [4 /*yield*/, file_1.default.section.readLast(repoPath)];
                    case 2: return [4 /*yield*/, _b.apply(_a, [(_c.body = _d.sent(),
                                _c)])];
                    case 3:
                        release = _d.sent();
                        if (config_1.default.release.github.open) {
                            opn(release.data.html_url, { wait: false });
                        }
                        if (!config_1.default.release.github.files.length) return [3 /*break*/, 8];
                        return [4 /*yield*/, globby(config_1.default.release.github.files, { cwd: cwd, absolute: true })];
                    case 4:
                        filePaths = _d.sent();
                        if (!filePaths.length) return [3 /*break*/, 8];
                        utils_1.default.log("Uploading " + filePaths.length + " files...");
                        _i = 0, filePaths_1 = filePaths;
                        _d.label = 5;
                    case 5:
                        if (!(_i < filePaths_1.length)) return [3 /*break*/, 8];
                        filePath = filePaths_1[_i];
                        utils_1.default.log("Uploading \"" + chalk_1.default.bold(filePath) + "\"");
                        return [4 /*yield*/, github.repos.uploadReleaseAsset({
                                url: release.data.upload_url,
                                name: path.basename(filePath),
                                file: fs.createReadStream(filePath),
                                headers: {
                                    'content-type': mime.lookup(filePath),
                                    'content-length': fs.statSync(filePath).size
                                }
                            })];
                    case 6:
                        _d.sent();
                        _d.label = 7;
                    case 7:
                        _i++;
                        return [3 /*break*/, 5];
                    case 8: return [3 /*break*/, 10];
                    case 9:
                        e_1 = _d.sent();
                        utils_1.default.log(e_1);
                        utils_1.default.exit('[release] An error occurred while making the GitHub release');
                        return [3 /*break*/, 10];
                    case 10: return [2 /*return*/];
                }
            });
        });
    }
};
/* EXPORT */
exports.default = GitHub;
