// @fabiospampinato/bump/dist/config.js
"use strict";
/* IMPORT */
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var minimist = require("minimist");
var os = require("os");
var path = require("path");
var utils_1 = require("./utils");
/* CONFIG */
var Config = {
    force: false,
    silent: false,
    files: {},
    version: {
        enabled: true,
        initial: '0.0.0',
        increments: ['major', 'minor', 'patch', 'premajor', 'preminor', 'prepatch', 'prerelease', 'custom'] // List of available increments to pick from
    },
    changelog: {
        enabled: true,
        create: false,
        open: true,
        file: 'CHANGELOG.md',
        version: '### Version [version]',
        commit: '- [message]',
        separator: '\n' // Template for the separator between versions sections
    },
    commit: {
        enabled: true,
        message: 'Bumped version to [version]' // Template for the commit message
    },
    tag: {
        enabled: true,
        name: 'v[version]' // Template for the name of the tag
    },
    release: {
        enabled: false,
        github: {
            enabled: false,
            open: true,
            draft: true,
            prerelease: false,
            files: [],
            token: '',
            owner: '',
            repo: '' // GitHub repository name
        }
    },
    tokens: {
        date: {
            format: 'YYYY-MM-DD' // Moment.js format to use when generating the `[date]` token
        },
        version_date: {
            format: 'YYYY-MM-DD' // Moment.js format to use when generating the `[version_date]` token
        }
    },
    scripts: {
        enabled: true,
        prebump: '',
        postbump: '',
        prechangelog: '',
        postchangelog: '',
        precommit: '',
        postcommit: '',
        pretag: '',
        posttag: '',
        prerelease: '',
        postrelease: '' // Script to execute after releasing
    }
};
/* LOCAL */
function initLocal() {
    var localPath = path.join(os.homedir(), '.bump.json'), localConfig = _.attempt(require, localPath);
    if (_.isError(localConfig))
        return;
    utils_1.default.config.merge(Config, localConfig);
}
initLocal();
/* CWD */
function initCwd() {
    var cwdPath = path.join(process.cwd(), 'bump.json'), cwdConfig = _.attempt(require, cwdPath);
    if (_.isError(cwdConfig))
        return;
    utils_1.default.config.merge(Config, cwdConfig);
}
initCwd();
/* DYNAMIC */
function initDynamic() {
    var argv = minimist(process.argv.slice(2));
    /* CONFIG */
    var dynamicPath = [argv.config, argv.c].find(_.isString), dynamicConfigRequire = _.attempt(require, path.resolve(process.cwd(), dynamicPath || '')), dynamicConfigJSON = _.attempt(JSON.parse, dynamicPath), dynamicConfig = !_.isError(dynamicConfigRequire) ? dynamicConfigRequire : (!_.isError(dynamicConfigJSON) ? dynamicConfigJSON : undefined);
    if (dynamicConfig)
        utils_1.default.config.merge(Config, dynamicConfig);
    /* COMMIT & TAG */
    if (!Config.commit.enabled)
        Config.tag.enabled = false;
    /* SWITCHES */
    var switches = ['silent', 'force'];
    switches.forEach(function (name) {
        Config[name] = [argv[name], Config[name]].find(_.isBoolean);
    });
    /* SCRIPTS */
    Config.scripts.enabled = [argv.scripts, Config.scripts.enabled].find(_.isBoolean);
    var scripts = Object.keys(Config.scripts);
    scripts.forEach(function (name) {
        if (!_.isString(Config.scripts[name]))
            return;
        Config.scripts[name] = [argv[name], Config.scripts[name]].find(_.isString);
    });
}
initDynamic();
/* INIT ENVIRONMENT */
function initEnvironment() {
    if (process.env.GITHUB_TOKEN)
        Config.release.github.token = process.env.GITHUB_TOKEN;
}
initEnvironment();
/* EXPORT */
exports.default = Config;
