// @fabiospampinato/bump/dist/providers/changelog/file.js
"use strict";
/* IMPORT */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
var moment = require("moment");
var opn = require("opn");
var path = require("path");
var inquirer_helpers_1 = require("inquirer-helpers");
var stringMatches = require("string-matches");
var config_1 = require("../../config");
var utils_1 = require("../../utils");
/* CHANGELOG */
var Changelog = {
    open: function (repoPath) {
        return __awaiter(this, void 0, void 0, function () {
            var changelogPath;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        changelogPath = path.join(repoPath, config_1.default.changelog.file);
                        return [4 /*yield*/, utils_1.default.file.exists(changelogPath)];
                    case 1:
                        if (!(_a.sent()))
                            return [2 /*return*/];
                        opn(changelogPath, { wait: false });
                        return [2 /*return*/];
                }
            });
        });
    },
    init: function (repoPath) {
        return __awaiter(this, void 0, void 0, function () {
            var bumps, sections, content;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, utils_1.default.repository.getVersionProvidersResult(repoPath, 'getCommitsBumps')];
                    case 1:
                        bumps = _a.sent();
                        if (!bumps.length)
                            return [2 /*return*/, utils_1.default.exit('[changelog] No commits found')];
                        sections = bumps.map(function (bump) { return Changelog.section.render(bump.version, bump.commits); }), content = sections.join('');
                        return [4 /*yield*/, Changelog.write(repoPath, content)];
                    case 2: return [2 /*return*/, _a.sent()];
                }
            });
        });
    },
    read: function (repoPath) {
        return __awaiter(this, void 0, void 0, function () {
            var changelogPath, content;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        changelogPath = path.join(repoPath, config_1.default.changelog.file);
                        return [4 /*yield*/, utils_1.default.file.read(changelogPath)];
                    case 1:
                        content = _a.sent();
                        return [2 /*return*/, content];
                }
            });
        });
    },
    update: function (repoPath, version) {
        return __awaiter(this, void 0, void 0, function () {
            var changelogPath, _a, bumps, section;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        changelogPath = path.join(repoPath, config_1.default.changelog.file);
                        return [4 /*yield*/, utils_1.default.file.exists(changelogPath)];
                    case 1:
                        if (!!(_b.sent())) return [3 /*break*/, 6];
                        if (!config_1.default.changelog.create && config_1.default.force)
                            return [2 /*return*/];
                        _a = config_1.default.changelog.create;
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, inquirer_helpers_1.default.noYes('No changelog found, do you want to create it?')];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        if (!_a) return [3 /*break*/, 5];
                        return [4 /*yield*/, Changelog.init(repoPath)];
                    case 4:
                        _b.sent();
                        _b.label = 5;
                    case 5: return [3 /*break*/, 8];
                    case 6: return [4 /*yield*/, utils_1.default.repository.getVersionProvidersResult(repoPath, 'getCommitsBumps', 1)];
                    case 7:
                        bumps = _b.sent();
                        if (!bumps.length)
                            return [2 /*return*/];
                        section = Changelog.section.render(version, bumps[0].commits);
                        return [2 /*return*/, Changelog.section.add(repoPath, section)];
                    case 8: return [2 /*return*/];
                }
            });
        });
    },
    write: function (repoPath, content) {
        return __awaiter(this, void 0, void 0, function () {
            var changelogPath, contentCleaned;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        changelogPath = path.join(repoPath, config_1.default.changelog.file), contentCleaned = content.replace(/^(\s*\r?\n){2,}/gm, '\n').replace(/(\s*\r?\n){2,}$/gm, '\n');
                        return [4 /*yield*/, utils_1.default.file.make(changelogPath, contentCleaned)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    },
    section: {
        readLast: function (repoPath) {
            return __awaiter(this, void 0, void 0, function () {
                var content, headerRe, matches;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, Changelog.read(repoPath)];
                        case 1:
                            content = _a.sent();
                            if (!content)
                                return [2 /*return*/];
                            headerRe = new RegExp("^" + _.escapeRegExp(config_1.default.changelog.version).replace(/\\\[[^\]]+\\\]/gi, '(.*)') + "$", 'gmi'), matches = stringMatches(content, headerRe, 2);
                            if (matches.length < 2)
                                return [2 /*return*/];
                            return [2 /*return*/, _.trim(content.slice(matches[0].index + matches[0][0].length, matches[1].index))];
                    }
                });
            });
        },
        render: function (version, commits) {
            /* VARIABLES */
            var lines = [];
            /* TOKENS */
            var tokens = {
                version: version,
                version_date: moment().format(config_1.default.tokens.version_date.format)
            };
            /* VERSION */
            if (config_1.default.changelog.version) {
                lines.push(utils_1.default.template.render(config_1.default.changelog.version, tokens));
            }
            /* COMMITS */
            if (config_1.default.changelog.commit) {
                commits.reverse().forEach(function (commit) {
                    if (commit.isBump)
                        return;
                    var mergeRe = /^(Merge PR|Merge pull request|Merge branch) /gi;
                    if (mergeRe.test(commit.message))
                        return;
                    var hash = commit.hash, date = commit.date, message = commit.message, author_name = commit.author_name, author_email = commit.author_email, messageCleaned = message.replace(/ \(HEAD\)$/i, '').replace(/ \(HEAD -> [^)]+\)$/i, '').replace(/ \(tag: [^)]+\)$/i, '');
                    var commitTokens = _.extend({}, tokens, {
                        date: moment(date).format(config_1.default.tokens.date.format),
                        message: messageCleaned,
                        hash: hash,
                        hash4: hash.slice(0, 4),
                        hash7: hash.slice(0, 7),
                        hash8: hash.slice(0, 8),
                        author_name: author_name,
                        author_email: author_email
                    });
                    lines.push(utils_1.default.template.render(config_1.default.changelog.commit, commitTokens));
                });
            }
            /* SECTION */
            var section = lines.join('\n') + '\n';
            /* SEPARATOR */
            if (config_1.default.changelog.separator) {
                section += utils_1.default.template.render(config_1.default.changelog.separator, tokens);
            }
            /* RETURN */
            return section;
        },
        add: function (repoPath, section) {
            return __awaiter(this, void 0, void 0, function () {
                var content;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0: return [4 /*yield*/, Changelog.read(repoPath)];
                        case 1:
                            content = _a.sent();
                            if (!content)
                                return [2 /*return*/, utils_1.default.exit('[changelog] Changelog missing')];
                            if (content.startsWith(section))
                                return [2 /*return*/]; // Duplicate section
                            content = section + content;
                            return [4 /*yield*/, Changelog.write(repoPath, content)];
                        case 2:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            });
        }
    }
};
/* EXPORT */
exports.default = Changelog;
